File: simple_implicit/driver.cpp
/* Main routine to test the backward Euler, trapezoidal, and
   forward Euler methods on the scalar-valued ODE problem
     y' = lambda*y + (1-lambda)*cos(t) - (1+lambda)*sin(t), t in [0,5],
     y(0) = 1.

   D.R. Reynolds
   Math 6321 @ SMU
   Fall 2020  */

#include <iostream>
#include "bwd_euler.hpp"
#include "fwd_euler.hpp"

using namespace std;
using namespace arma;


// Define classes to compute the ODE RHS function and its Jacobian

//    ODE RHS function class -- instantiates a RHSFunction
class MyRHS: public RHSFunction {
public:
  double lambda;                              // stores some local data
  int Evaluate(double t, vec& y, vec& f) {    // evaluates the RHS function, f(t,y)
    f(0) = lambda*y(0) + (1.0-lambda)*cos(t) - (1.0+lambda)*sin(t);
    return 0;
  }
};

//    ODE RHS Jacobian function class -- instantiates a RHSJacobian
class MyJac: public RHSJacobian {
public:
  double lambda;                              // stores some local data
  int Evaluate(double t, vec& y, mat& J) {    // evaluates the RHS Jacobian, J(t,y)
    J(0,0) = lambda;
    return 0;
  }
};


// Convenience function for analytical solution
vec ytrue(const double t) {
  vec yt(1);
  yt(0) = sin(t) + cos(t);
  return yt;
};



// main routine
int main() {

  // time steps to try
  vec h("1.0, 0.1, 0.01, 0.001");

  // lambda values to try
  vec lambdas("-1.0, -10.0, -50.0");

  // set problem information
  vec y0("1.0");
  double t0 = 0.0;
  double Tf = 5.0;

  // set desired output times
  int Nout = 6;  // includes initial condition
  vec tspan = linspace(t0, Tf, Nout);

  // create ODE RHS and Jacobian objects
  MyRHS rhs;
  MyJac Jac;

  // create true solution results
  mat Ytrue(1,Nout);
  for (size_t i=0; i<Nout; i++) {
    Ytrue.col(i) = ytrue(tspan(i));
  }

  // create time stepper objects
  BackwardEulerStepper BE(rhs, Jac, y0);
  ForwardEulerStepper  FE(rhs, y0);

  // update Newton solver parameters
  BE.newt.tol = 1.e-3;
  BE.newt.maxit = 20;
  BE.newt.show_iterates = false;

  //------ Backwards Euler tests ------

  // loop over lambda values
  for (int il=0; il<lambdas.n_elem; il++) {

    // set current lambda value into rhs and Jac objects
    rhs.lambda = lambdas(il);
    Jac.lambda = lambdas(il);

    // loop over time step sizes
    for (int ih=0; ih<h.n_elem; ih++) {

      // call stepper
      cout << "\nRunning backward Euler with stepsize h = " << h(ih)
	   << ", lambda = " << lambdas(il) << ":\n";
      mat Y = BE.Evolve(tspan, h(ih), y0);

      // output solution, errors, and overall error
      mat Yerr = abs(Y-Ytrue);
      for (size_t i=0; i<Nout; i++)
        printf("    y(%.1f) = %9.6f   |error| = %.2e\n",
               tspan(i), Y(0,i), Yerr(0,i));
      cout << "  Max error = " << Yerr.max() << endl;

    }
    cout << endl;
  }


  //------ Forward Euler tests ------

  // loop over lambda values
  for (int il=0; il<lambdas.n_elem; il++) {

    // set current lambda value into rhs and Jac objects
    rhs.lambda = lambdas(il);
    Jac.lambda = lambdas(il);

    // loop over time step sizes
    for (int ih=0; ih<h.n_elem; ih++) {

      // call stepper
      cout << "\nRunning forward Euler with stepsize h = " << h(ih)
	   << ", lambda = " << lambdas(il) << ":\n";
      mat Y = FE.Evolve(tspan, h(ih), y0);

      // output solution, errors, and overall error
      mat Yerr = abs(Y-Ytrue);
      for (size_t i=0; i<Nout; i++)
        printf("    y(%.1f) = %9.6f   |error| = %.2e\n",
               tspan(i), Y(0,i), Yerr(0,i));
      cout << "  Max error = " << Yerr.max() << endl;

    }
    cout << endl;
  }

  return 0;
}

File: simple_implicit/driver.py
#!/usr/bin/env python3
#
# Main routine to test the backward Euler, trapezoidal, and
# forward Euler methods on the scalar-valued ODE problem
#    y' = lambda*y + (1-lambda)*cos(t) - (1+lambda)*sin(t), t in [0,5],
#    y(0) = 1.
#
# D.R. Reynolds
# Math 6321 @ SMU
# Fall 2023
import numpy as np
import sys
sys.path.append('..')
from termcolor import colored
from shared.ImplicitSolver import *
from forward_euler.ForwardEuler import *
from BackwardEuler import *
from Trapezoidal import *

# problem time interval and parameters
t0 = 0.0
tf = 5.0

# problem-defining functions
def ytrue(t):
    """ Generates a numpy array containing the true solution to the IVP at a given input t. """
    return np.array([np.sin(t) + np.cos(t)])
def f(t,y,lam):
    """ Right-hand side function, f(t,y), for the IVP """
    return np.array([lam*y[0] + (1.0-lam)*np.cos(t) - (1.0+lam)*np.sin(t)])
def J(t,y,lam):
    """ Jacobian (in dense matrix format) of the right-hand side function, J(t,y) = df/dy """
    return np.array( [ [lam] ] )

# construct implicit solvers
solver = ImplicitSolver(J, solver_type='dense', maxiter=20, rtol=1e-9, atol=1e-12, Jfreq=2)

# shared testing data
Nout = 6   # includes initial condition
tspan = np.linspace(t0, tf, Nout)
Ytrue = np.zeros((Nout, 1))
for i in range(Nout):
    Ytrue[i,:] = ytrue(tspan[i])
y0 = Ytrue[0,:]
hvals = np.array([1.0, 0.1, 0.01, 0.001])
errs = np.zeros(hvals.size)

# create forward Euler, backward Euler, and trapezoidal solvers
BE = BackwardEuler(f, solver)
Tr = Trapezoidal(f, solver)
FE = ForwardEuler(f)


# loop over stiffness values
for lam in [-1.0, -10.0, -50.0]:

    # backward Euler tests
    print(colored("\nbackward Euler tests:", "yellow", attrs=["bold"]))
    for idx, h in enumerate(hvals):
        print("  h = ",h,",  lambda = ", lam,":", sep='')
        BE.reset()
        BE.sol.reset()
        # Note that this is where we provide the rhs function parameter lam -- the "," is
        # required to ensure that args is an iterable (and not a float).
        Y, success = BE.Evolve(tspan, y0, h, args=(lam,))
        Yerr = np.abs(Y-Ytrue)
        errs[idx] = np.linalg.norm(Yerr,np.inf)
        if (success):
            print("     " + colored("  t      y(t)     |err(t)| ", attrs=["underline"]))
            for i in range(Nout):
                text = "      %.1f  %10.2e  %.2e" % (tspan[i], Y[i,0], Yerr[i,0])
                if (Yerr[i,0] > 1):
                    print(colored(text, "light_red"))
                else:
                    print(text)
            text = "  overall:  steps = %4i  Niters = %6i  NJevals = %5i  abserr = %8.2e" \
                % (BE.get_num_steps(), BE.sol.get_total_iters(),
                   BE.sol.get_total_setups(), errs[idx])
            if (errs[idx] > 1):
                print(colored(text, "red"))
            else:
                print(colored(text, "green"))
    orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
    print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))

    # trapezoidal tests
    print(colored("\ntrapezoidal tests:", "yellow", attrs=["bold"]))
    for idx, h in enumerate(hvals):
        print("  h = ",h,",  lambda = ", lam,":", sep='')
        Tr.reset()
        Tr.sol.reset()
        Y, success = Tr.Evolve(tspan, y0, h, args=(lam,))
        Yerr = np.abs(Y-Ytrue)
        errs[idx] = np.linalg.norm(Yerr,np.inf)
        if (success):
            print("     " + colored("  t      y(t)     |err(t)| ", attrs=["underline"]))
            for i in range(Nout):
                text = "      %.1f  %10.2e  %.2e" % (tspan[i], Y[i,0], Yerr[i,0])
                if (Yerr[i,0] > 1):
                    print(colored(text, "light_red"))
                else:
                    print(text)
            text = "  overall:  steps = %4i  Niters = %6i  NJevals = %5i  abserr = %8.2e" % \
                  (Tr.get_num_steps(), Tr.sol.get_total_iters(),
                   Tr.sol.get_total_setups(), errs[idx])
            if (errs[idx] > 1):
                print(colored(text, "red"))
            else:
                print(colored(text, "green"))
    orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
    print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))

    # forward Euler tests
    print(colored("\nforward Euler tests:", "yellow", attrs=["bold"]))
    for idx, h in enumerate(hvals):
        print("  h = ",h,",  lambda = ", lam,":", sep='')
        FE.reset()
        Y, success = FE.Evolve(tspan, y0, h, args=(lam,))
        Yerr = np.abs(Y-Ytrue)
        errs[idx] = np.linalg.norm(Yerr,np.inf)
        if (success):
            print("     " + colored("  t      y(t)     |err(t)| ", attrs=["underline"]))
            for i in range(Nout):
                text = "      %.1f  %10.2e  %.2e" % (tspan[i], Y[i,0], Yerr[i,0])
                if (Yerr[i,0] > 1):
                    print(colored(text, "light_red"))
                else:
                    print(text)
            text = "  overall:  steps = %4i  abserr = %8.2e" % (FE.get_num_steps(), errs[idx])
            if (errs[idx] > 1):
                print(colored(text, "red"))
            else:
                print(colored(text, "green"))
    orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
    print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))
